core/groupby.cpp

```diff
@@ -60,24 +60,32 @@ template <class T> T GroupBy<T>::flatten(const Dim reductionDim) const {
   return out;
 }
 
+template <class T>
+void sum_impl(const VariableProxy &out_data, const T &data_container,
+              const std::vector<Slice> &group, const Dim reductionDim,
+              const MasksConstProxy &masks) {
+  for (const auto &slice : group) {
+    const auto merged_inverted_masks = ~masks_merge(masks, reductionDim, slice);
+    sum_impl(out_data,
+             data_container.data().slice(slice) * merged_inverted_masks);
+  }
+}
+
 /// Apply sum to groups and return combined data.
 template <class T> T GroupBy<T>::sum(const Dim reductionDim) const {
   auto out = makeReductionOutput(reductionDim);
   // Apply to each group, storing result in output slice
   for (scipp::index group = 0; group < size(); ++group) {
     const auto out_slice = out.slice({dim(), group});
+    const auto masks = m_data.masks();
     if constexpr (std::is_same_v<T, Dataset>) {
       for (const auto &[name, item] : m_data) {
         const auto out_data = out_slice[name].data();
-        for (const auto &slice : groups()[group]) {
-          sum_impl(out_data, item.data().slice(slice));
-        }
+        sum_impl(out_data, item, groups()[group], reductionDim, masks);
```

Comment:
No need for the `masks` argument, `item` should contain all relevant masks anyway?